W-11862525-first-draft-for-beta-site-build-CP (#114)
mulesoft/docs-apikit: modules/ROOT/pages/creating-an-odatav4-api-with-apikit.adoc

= Building, Implementing, and Testing an OData V4 API
ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]
:page-aliases: creating-an-odata-api-with-apikit.adoc

You can query a MySQL database with the xref:creating-an-odatav4-api-with-apikit.adoc#apikit-for-odata-4-implementation-example[OData MySQL example] by using an HTTP-based data service.

== Before You Begin

You must install the following software to create and use an OData REST API with APIkit:

* OData Plugin
* Mule runtime engine 4.3.0 or later
* Anypoint Studio 7.9.0 or later

Additionally, you must download the link:{attachmentsdir}/apikit-odata4-example.jar[OData 4 implementation example].

== Scaffold an OData V4 API Using APIkit for OData

. In Studio, select *File* > *New* > *Mule Project*.
. In the *Project Name* field, type the name of your Mule project.
. Click *Finish*.

> . In `+src/main/resources/api+`, create a new file with a `.csdl.xml` extension. +

Reviewer: Global - put the + on a line by itself. This affects the spacing. See https://confluence.internal.salesforce.com/display/MTDT/Formatting+with+Asciidoc#FormattingwithAsciidoc-ListContinuation(akaIndentedParagraphs)

Author: OK! Thanks!

  </edmx:DataServices>
</edmx:Edmx>
--
[calloutlist]
.. `EntityType` defines the type and its properties.
.. `key` defines which property or set of properties is the key for the entity. +

> For example, the entity set `Orders` is composed of two keys: `OrderID` and `ShipName`.

Reviewer: ```suggestion
For example, the entity set `Orders` comprises two keys: `OrderID` and `ShipName`.
```

Author: OK! Thanks!

}'
--

> OData V4 discourages the usage of PUT to update your entities because it replaces the entire entity with a new one and this operation can lead to data loss.

Reviewer: ```suggestion
OData V4 discourages the usage of PUT to update your entities because it replaces the entire entity with the new one, which can lead to data loss.
```

Author: OK! Thanks!

}'
--

> OData V4 discourages the usage of PUT to update your entities because it replaces the entire entity with the new one, which can lead to data loss.

Reviewer 2: ```suggestion
OData V4 discourages the usage of PUT to update your entities because PUT replaces the entire entity with the new one, which can lead to data loss.
```

Author: OK! Thanks!

=== Server-Side Pagination in APIkit for OData 4

> Server-side pagination is a way of controlling a subset of data requests fetched from a client-side server by using a fixed `pageSize` parameter. It divides the response data set into discrete pages to improve data readability. The server defines the page size, which is the maximum number of records the server returns in a request.

Reviewer 2: ```suggestion
Server-side pagination enables you to control a subset of data requests fetched from a client-side server by using a fixed `pageSize` parameter. It divides the response data set into discrete pages to improve data readability. The server defines the page size, which is the maximum number of records that the server returns in a request.
```

==== Server-Side Pagination Operation in OData 4

OData defines a `@odata.nextLink` to indicate that a response is only a subset of the requested collection of entities. As such, `@odata.nextLink` contains a URL that enables the client to retrieve the next subset of the requested collection. The URL contains a query parameter `$skiptoken` that indicates the server where the next page starts.

==== Implement Server-Side Pagination in APIkit for OData 4

You can implement the server-side pagination by defining a set of fields with `<apikit-odata:request-entity-collection-listener>` in the *Responses* tab in Anypoint Studio.

image::pagesize.png["pageSize set to ${service.orders.pageSize} in the On Entity Collection Request tab"]

> After it is populated, APIkit creates the XML representation of the previous fields:

Reviewer 2: What does "it" refer to?

Author: Requesting assistance from SMEs.

Author: It refers to the `pageSize`.
